Fix playing videos in the radar widget

penev92 opened the pull request:
Fixes #20622.

The original crash that was reported was because `Delegate.BeginInvoke()` is not supported on .NET Core and later, which I mention on the issue.

Diff probably looks scarier than it is.
~~This unified approach for playing videos from Lua removes the "async loading" from the radar videos (that wasn't used by the fullscreen videos!!), but I couldn't feel any severe lag so I hope that is fine.~~

penev92 committed:
Did a slight refactor pass on MediaGlobal

- Unified onPlayComplete callback handling to avoid code duplication.
- Fixed code style issues.

penev92 committed:
Unify MediaGlobal/Scripting.Media PlayFMV* methods

- Improve consistency
- Remove dead code
- Removed AsyncAction!
- Delegate.BeginInvoke is unsupported since .NET Core and throws an "Operation is not supported on this platform." exception.

penev92 committed:
Rename some VideoPlayerWidget methods for clarity

PunkPun reviewed (commented):
You have to have the videos installed for them to play correct? I don't have music / videos imported on my client and it didn't crash / do anything other than display messages in chat, so I guess it's working as intended.

But I'm concerned about the removal of async loading. IMO we should preserve and fix it, though for prep I think it's perfectly fine if we dropped it. 

Other than 👍

penev92 commented:
I don't think it's a big issue and I don't think we do it anywhere else (so no precedent), but I am open to suggestions about how to do it. It should be much lower-level than the Lua API though and ideally used for all assets everywhere.

abcdefg30 reviewed (dismissed):
Not convinced that not drawing the overlay is a "fix", but this is good enough for prep. Thanks for tracking this down/completing the fix.

penev92 committed:
Added VideoPlayerWidget.LoadAndPlayAsync method

Previous commits removed the async loading of videos, which can be a problem for videos played in the radar widget mid-game because it can cause a lag spike. This loads the video on a separate thread and runs it on the main thread whenever it is loaded, not blocking the main thread meanwhile and allowing the game to continue while the video loads.
Also add back cancelling of already playing video and add a check to not try to run onComplete if it is null.

penev92 commented:
> Not convinced that not drawing the overlay is a "fix", but this is good enough for prep.

Not drawing the overlay is just ***"a"*** fix to a problem that came up while working on this. It has nothing to do with the original reported crash.

In any case I added one more point to the commit's description to make it clearer what it does and I added another commit adding "back" async video loading so we don't risk any delays when loading a video mid-game. We may actually want to consider switching fullscreen video playing to use the async method as well (but that is also used by victory/loss videos making it a bit more iffy).
I guess the new commit invalidates @abcdefg30's 👍  😕 

P.S.: Also added more test cases.

penev92 commented:
Rebased and removed `AsyncAction` as its only use-case was removed.

PunkPun commented:
Fullscreen video executes OnComplete upon a failed video load attempt, while radar video doesn't

penev92 commented:
Fixed. Now they both call the callback function if the file is missing (we don't handle any other exceptions in either method, but that looks like it is by design).

PunkPun commented on OpenRA.Mods.Common/Scripting/Media.cs:
```diff
@@ -60,27 +59,11 @@ public static void PlayFMVFullscreen(World w, string movie, Action onComplete)
-		public static IVideo LoadVideo(Stream s)
-		{
-			return VideoLoader.GetVideo(s, true, Game.ModData.VideoLoaders);
+			player.DrawOverlay = false;
```
I still don't understand this. It's never set to true, why is it here? Why is it needed? An explanation should be added to the variable

penev92 replied:
The default value is `true`: https://github.com/OpenRA/OpenRA/blob/5bf7fe852c8f122e3a3e0a1eb72d66d7429675c3/OpenRA.Mods.Common/Widgets/VideoPlayerWidget.cs#L24

I guess there was another crash hiding behind the `.BeginInvoke()` one, so I added that line on a hunch (which paid off). Investigating it further now, I guess the issue is related to this:
![video_played_widget_draw_overlay](https://user-images.githubusercontent.com/7137365/217160511-5d5fc81b-e913-4a0c-9c2a-a91cf6ad9b0f.gif)
I have a better fix that I will push tonight after work.

penev92 committed:
Fix crash caused by VideoPlayerWidget.DrawOverlay

Somehow this would divide by 0 on the next line, not crashing but resulting in an integer overflow on `overlayHeight`, which would crash when trying to render.

penev92 commented:
Updated.

penev92 commented:
Added back video cancelling by another video in the last commit.

penev92 committed:
Added video player background for D2k

penev92 commented:
Also added a race condition fix for when loading 2 videos for the radar widget at the same time
and snuck in the addition of a full-screen in-gave video player background for D2k as extra polish.

PunkPun reviewed (dismissed):
LGTM

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/f9c71461ec340cc4af6e10cf21fff37ae32be387...375e776baf215f68e52a1c9bce97c3f8e8ab02d3)

https://github.com/OpenRA/OpenRA/commit/44f35729e117c0476071b50002648c5b9cc99aff
https://github.com/OpenRA/OpenRA/commit/d7b2727d16a147828f544b7ce3799377c050ab0f
https://github.com/OpenRA/OpenRA/commit/d99991d317ec2c95d3802555e095fea103c3480f
https://github.com/OpenRA/OpenRA/commit/1f63294503bae868650386c679a424cc66ddb6df
https://github.com/OpenRA/OpenRA/commit/7bc2768dc4cea18f7ca8b1013232bde7d80af6ae
https://github.com/OpenRA/OpenRA/commit/7867eb99c6e48498ac61a3f7cc3737c1fded7728
